scrapy/extensions/robotstxtparser.py

```diff
@@ -0,0 +1,52 @@
+from six.moves.urllib_robotparser import RobotFileParser
+
+from scrapy.utils.python import to_native_str
+
+class BaseRobotsTxtParser():
+    def __init__(self, robotstxt_url):
+        raise NotImplementedError
+
+    def parse(self, robotstxt_body):
+        """Parse content of robots.txt file."""
+        raise NotImplementedError
+
+    def allowed(self, url, useragent):
+        """Return True if url is allowed for crawling, otherwise return False."""
+        raise NotImplementedError
+
+    def sitemaps(self):
+        """Return a list of links of sitemaps on the website. If there is no sitemap specified in robots.txt, return None."""
+        raise NotImplementedError
+
+    def crawl_delay(self, useragent):
+        """Return time specified with Crawl-delay directive. If nothing is specified, return None."""
+        raise NotImplementedError
+
+    def preferred_host(self):
+        """Return preferred domain specified with Host directive. If nothing is specified, return None.""" 
+        raise NotImplementedError
+
+class PythonRobotParser(BaseRobotsTxtParser):
+    def __init__(self, robotstxt_url):
+        self.robotstxt_url = robotstxt_url
+        self.rp = RobotFileParser(robotstxt_url) 
+
+    def parse(self, robotstxt_body):
+        self.rp.parse(to_native_str(robotstxt_body).splitlines())
```

Comment:
Convert to proper type of passed data unsing `to_native_str` should be done once on a side of `RobotsTxtMiddleware`. Otherwise every implementation have to do same thing

Reply:
@anubhavp28 has actually done it on the `RobotFileParser` implementation, not in the middleware, but I believe that is the right approach here.

I was not familiar with `to_native_str`, but it turns out it returns bytes in Python 2 and a string in Python 3 to deal with those Python 2 implementations that did not support Python 2’s `unicode` type, such as `RobotFileParser`.

Other parsers might need to use bytes on both versions of Python or a string on both versions of Python, so it makes sense that `to_native_str` is used by the implementations that need it and not forced by the middleware.